fuzz: wallet: add target for spkm migration

brunoerg opened the pull request:
This PR adds a fuzz target for `ScriptPubKeyMan` migration. It creates a `LegacyDataSPKM` which can have some keys/scripts/etc, and then migrate it to descriptor. I tried to keep it as compatible as possible with future legacy wallet removal.

brunoerg committed:
fuzz: wallet: add target for spkm migration

brunoerg commented:
cc: @achow101

furszy commented on src/wallet/test/fuzz/scriptpubkeyman.cpp:
```diff
@@ -196,5 +203,77 @@ FUZZ_TARGET(scriptpubkeyman, .init = initialize_spkm)
+
+    if (keys.empty()) return;
+
+    auto& legacy_spkm{*wallet.GetOrCreateLegacyScriptPubKeyMan()};
```
The legacy spkm class will be deleted soon in #28710.
You might need to add a mechanism to mock the bdb reader class just so you can feed the migration process with a hand-crafted `LegacyDataSPKM`. This will let you avoid crafting the bdb database file manually when bdb is not compiled anymore.

brunoerg replied:
> You might need to add a mechanism to mock the bdb reader class just so you can feed the migration process with a hand-crafted LegacyDataSPKM. This will let you avoid crafting the bdb file manually when bdb is not compiled anymore.

You're right, I'll move this PR to draft to work on it.

brunoerg replied:
Just changed it to use `LegacyDataSPKM`. Note that this target is about the spkm migration only, not the whole process.

brunoerg commented:
Inputs for this are available at: https://github.com/brunoerg/qa-assets/pull/2

achow101 commented on src/wallet/test/fuzz/scriptpubkeyman.cpp:
```diff
@@ -200,5 +207,88 @@ FUZZ_TARGET(scriptpubkeyman, .init = initialize_spkm)
+                    load_key_count++;
+                    if (fuzzed_data_provider.ConsumeBool()) {
+                        CHDChain hd_chain;
+                        hd_chain.nVersion = CHDChain::VERSION_HD_CHAIN_SPLIT;
```
I think it would also be good to cover hd chain before hd chain split.

brunoerg replied:
Will do it.

achow101 commented on src/wallet/test/fuzz/scriptpubkeyman.cpp:
```diff
@@ -200,5 +207,88 @@ FUZZ_TARGET(scriptpubkeyman, .init = initialize_spkm)
+    std::optional<MigrationData> res{legacy_data.MigrateToDescriptor()};
+    if (res.has_value()) {
+        assert(static_cast<int>(res->desc_spkms.size()) >= load_key_count);
+        if (!res->solvable_descs.empty()) assert(script_count > 0);
```
Would it be possible for `script_count` to really be watchonly script count and just have it increment only when a watchonly script is added?

brunoerg replied:
I think so, it just needs to check if the script was built with a loaded key.

brunoerg replied:
In fact, I tried to track watchonly and spendable scripts, but I don't think it's simple since we load keys, hd chain, diverse scripts, etc. I ended up writing similar or copying a lot of things from `wallet/scriptpubkey.cpp` which I don't think would be a valuable approach here.

brunoerg replied:
Done.

achow101 commented on src/wallet/test/fuzz/scriptpubkeyman.cpp:
```diff
@@ -200,5 +207,88 @@ FUZZ_TARGET(scriptpubkeyman, .init = initialize_spkm)
+
+    std::optional<MigrationData> res{legacy_data.MigrateToDescriptor()};
+    if (res.has_value()) {
+        assert(static_cast<int>(res->desc_spkms.size()) >= load_key_count);
```
If watchonly script count can be tracked, then this check should be able to be an exact check as `desc_spkms` should be `load_key_count` plus the number of spendable scripts.

brunoerg replied:
Done.

achow101 commented on src/wallet/test/fuzz/scriptpubkeyman.cpp:
```diff
@@ -200,5 +207,88 @@ FUZZ_TARGET(scriptpubkeyman, .init = initialize_spkm)
+                CScript script{ConsumeScript(fuzzed_data_provider)};
+                const auto key_index{fuzzed_data_provider.ConsumeIntegralInRange<int>(0, keys.size() - 1)};
+                LIMITED_WHILE(fuzzed_data_provider.ConsumeBool(), 2) {
+                    CallOneOf(
```
It would be nice to have multisig scripts here as the handling of ismine for that is quite complicated and having fuzz testing of it would be great.

brunoerg replied:
Sure, I will do it.

brunoerg commented:
Force-pushed addressing (multisig and hd chain cover): 

- https://github.com/bitcoin/bitcoin/pull/29694#discussion_r1828420886
- https://github.com/bitcoin/bitcoin/pull/29694#discussion_r1828423452

furszy reviewed (commented):
Would be good to ensure that #31374 can be replicated here.

brunoerg commented:
Force-pushed addressing https://github.com/bitcoin/bitcoin/pull/29694#discussion_r1828422802

brunoerg commented:
> Would be good to ensure that https://github.com/bitcoin/bitcoin/pull/31374 can be replicated here.

Since this target is only for spkm migration, I don't think it can be replicated here.

furszy commented on src/wallet/test/fuzz/scriptpubkeyman.cpp:
```diff
@@ -200,5 +207,129 @@ FUZZ_TARGET(scriptpubkeyman, .init = initialize_spkm)
+    LOCK(wallet.cs_wallet);
+    CBlockLocator loc{chainstate.m_chain.GetLocator()};
+    assert(batch.WriteBestBlock(loc));
+    (void)wallet.ApplyMigrationData(batch, *res);
```
You cannot call `ApplyMigrationData` without executing the entire `DoMigration step`. Tat's just incorrect. This code skips the creation of the watch-only and solvable wallets from the `MigrationData` result, causing a crash inside `ApplyMigrationData`.

brunoerg replied:
Yes, just realized it. I'm going to change it, but afaik `DoMigration` can't be called directly here.

brunoerg replied:
Done

brunoerg commented:
Force-pushed:

Now the fuzz target runs the entire migration process (`MigrateLegacyToDescriptor`). Also, I added a new commit that allows to migrate in-memory wallets to speed up the target and avoid timeouts.

brunoerg committed:
wallet: add option to create in-memory SQLite DB in the migration

brunoerg committed:
wallet: sqlite: there is no need to have exclusive locking when mocking

For in-memory SQLite databases, exclusive locking is generally not
necessary because in-memory dbs are private to the connection that
created them.

brunoerg committed:
wallet: skip backup restoration when using in-memory dbs

By using in-memory databases, we do not have dbs file to delete
and restore.

brunoerg commented:
Force-pushed adding `SeedRandomStateForTest` and `SetMockTime` for better stability.

brunoerg commented:
Rebased

brunoerg commented:
Just fixed CI, ready for review!

pablomartin4btc reviewed (commented):
You are skipping the backup restore in commit: "_wallet: skip backup restoration when using in-memory dbs_"; shouldn't the backup itself be skipped too when `in_memory`?

brunoerg commented on src/wallet/wallet.cpp:
```diff
@@ -4551,11 +4552,14 @@ util::Result<MigrationResult> MigrateLegacyToDescriptor(std::shared_ptr<CWallet>
         }
     }
-    if (!success) {
+    if (!success && !in_memory) {
```
Yes, these other things are also not necessary when using in memory db. I'm gonna address it.

brunoerg replied:
Why? the for after `// Unload the wallets` are basically playing with `created_wallets` which is used only if the db is not in memory. Same for the lines after `// Verify that there is no dangling wallet` which needs `ptr_wallet`.

pablomartin4btc replied:
> the for after `// Unload the wallets`...

There's the removal of the wallet from the context (`if (!RemoveWallet(context, w, /*load_on_start=*/false)) {`) using `created_wallets` - that was done when the wallet was loaded (in `LoadWalletInternal`-> `AddWallet(context, wallet);` ).

> the lines after `// Verify that there is no dangling wallet` which needs `ptr_wallet`.

At the end of the condition is returning an error (`return util::Error{error};`)

Be aware that if you remove the commit [wallet: skip backup restoration when using in-memory dbs](https://github.com/bitcoin/bitcoin/pull/29694/commits/6f0e025efad60bceb2cce7a4db631126ac9b22cb) altogether you need to add a conditional to avoid the restore section.

At some point I think we need to refactor the wallet and extract the wallet migration behaviour into a new/ few object/ s.

brunoerg commented:
I'm reworking this target, will draft it.

brunoerg commented:
Closing in favor of #32624.
